cupy/core/_gufuncs.py

```diff
@@ -0,0 +1,500 @@
+        # so we can avoid most of the __call__ stuff
+        self._func = func
+        self._signature = signature
+        self._default_casting = 'same_kind'
+
+        # The following are attributes to avoid applying certain steps
+        # when wrapping cupy functions that do some of the gufunc
+        # stuff internally due to CUDA libraries requirements
+        self._supports_batched = kwargs.get('supports_batched', False)
+        self._supports_out = kwargs.get('supports_out', False)
+
+        # Preprocess the signature here
+        input_coredimss, output_coredimss = _parse_gufunc_signature(
+            self._signature)
+        self._input_coredimss = input_coredimss
+        self._output_coredimss = output_coredimss
+        # This is pre-calculated to later check the minimum number of
+        # dimensions required per input
+        self._min_dims = [0] * len(input_coredimss)
+        for i, inp in enumerate(input_coredimss):
+            for d in inp:
+                if d[-1] != '?':
+                    self._min_dims[i] += 1
+
+        # Determine nout: nout = None for functions of one
+        # direct return; nout = int for return tuples
+        self._nout = (
+            None
+            if not isinstance(output_coredimss, list)
+            else len(output_coredimss)
+        )
+
+    def _apply_func_to_inputs(self, dim, sizes, dims, args, outs):
+        # Apply function
+        # The resulting array is loop_output_dims+the specified dims
+        # Some functions have batching logic inside due to higly
+        # optimized CUDA libraries so we just call them
+        if self._supports_batched or dim == len(dims):
+            # Check if the function supports out, order and other args
+            if self._supports_out and outs is not None:
+                outs = outs[0] if len(outs) == 0 else outs
+                self._func(*args, out=outs)
+            else:
+                fouts = self._func(*args)
+                # TODO(ecastill) improve this check
+                if isinstance(fouts, cupy.ndarray):
+                    fouts = (fouts,)
+                for o, fo in zip(outs, fouts):
+                    o[...] = fo
+        else:
+            dim_size = sizes[dims[dim]][0]
+            for i in range(dim_size):
+                n_args = [a[i] for a in args]
+                if outs is not None:
+                    n_outs = [o[i] for o in outs]
+                    self._apply_func_to_inputs(
+                        dim + 1, sizes, dims, n_args, n_outs)
+
+    def _transpose_element(self, arg, iax, shape):
+        iax = tuple(a if a < 0 else a - len(shape) for a in iax)
+        tidc = (
+            tuple(i for i in range(
+                -len(shape) + 0, 0) if i not in iax) + iax
+        )
+        return arg.transpose(tidc)
+
+    def _get_args_transposed(self, args, input_axes, input_coredimss,
+                             outs, output_axes, output_coredimss):
```

Comment:
Could you remove `input_coredimss` and `output_coredimss` argument and use `self.*` instead?